src/Microsoft.Azure.SignalR.Management/HubInstanceFactories/ServiceHubLifetimeManagerFactory.cs

```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.ComponentModel;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.SignalR.Management
+{
+    internal class ServiceHubLifetimeManagerFactory
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly ServiceManagerContext _context;
+        public ServiceHubLifetimeManagerFactory(IServiceProvider sp, IOptions<ServiceManagerContext> context)
+        {
+            _serviceProvider = sp;
+            _context = context.Value;
+        }
+        public async Task<IServiceHubLifetimeManager> Create(string hubName, CancellationToken cancellationToken, ILoggerFactory loggerFactoryPerHub=null)
+        {
+            switch (_context.ServiceTransportType)
+            {
+                case ServiceTransportType.Persistent:
+                    {
+                        var containerFactory = _serviceProvider.GetRequiredService<MultiEndpointConnectionContainerFactory>();
+                        var container = containerFactory.Create(hubName,loggerFactoryPerHub);
+                        var connectionManager = new ServiceConnectionManager<Hub>();
+                        connectionManager.SetServiceConnection(container);
+                        _ = connectionManager.StartAsync();
+                        await container.ConnectionInitializedTask.OrTimeout(cancellationToken);
+                        return loggerFactoryPerHub == null ? ActivatorUtilities.CreateInstance<WebSocketsHubLifetimeManager<Hub>>(_serviceProvider, connectionManager) : ActivatorUtilities.CreateInstance<WebSocketsHubLifetimeManager<Hub>>(_serviceProvider, connectionManager, loggerFactoryPerHub);
```

Comment:
since you already put most of the class as a service into service collection, why not put WebSocketsHubLifetimeManager and RestHubLifetimeManager into service collection as well?

Reply:
`WebSocketsHubLifetimeManager` and `RestHubLifetimeManager` are not singletons, they are for hub. And we don't know the hub name until `CreateServiceHubContext(...)` is called, while hub name is bound to `ServiceHubLifetimeManager`.

Reply:
I thinks it's better prepare all the thing before actually start the container, so put the startAsync to `ServiceHubContextFactory.CreateAsync` is better

Reply:
What is the actual difference between starting connection manager in `ServiceHubLifetimeManagerFactory` and `ServiceHubContextFactory`? Now `ServiceHubContextFactory` doesn't need to know the actual impl of `IServiceHubLifetimeManager` . I think to put `connectionManager.StartAsync()` in `ServiceHubContextFactory` breaks the hierarchy of the code.

Reply:
what I mean is to put the used classes into service collections if possible, when everything is ready, start the connections

Reply:
This is not singleton and we can't create connectionManager until we know hub name?